src/api/plans.ts

```diff
@@ -96,19 +108,68 @@ router.delete("/plans/:planId", async (req, res) => {
   res.send();
 });
 
-router.get("/plans/:planId/courses", (req, res) => {
-  res.send({
-    courseName: "string",
-    courseNumber: 4056,
-    units: 4,
-    crosslisted: 53,
-    prerequisites: [
-      {
-        needed: 3,
-        "courses:": [12, 52, 236],
-      },
-    ],
-    offered: ["F", "W", "SP", "SU"],
-    description: "course description",
+router.get("/plans/:planId/courses", async (req, res) => {
+  const planCourses = await sql<PlanInfo[]>`
+    WITH recursive R AS(
+    SELECT program_courses.course_id as course FROM plans
+    JOIN programs on plans.program_id = programs.id
+    JOIN program_courses ON programs.id = program_courses.program_id
+    WHERE plans.id = ${req.params.planId}
+    UNION ALL
+    SELECT prerequisite_course_sets.course_id AS course FROM R
+    JOIN course_prerequisites ON course_prerequisites.course_id = course
+    JOIN prerequisite_course_sets ON course_prerequisites.set_number = prerequisite_course_sets.set_number 
+  )
+  SELECT Distinct course, course_prerequisites.set_number AS set, prerequisite_course_sets.course_id as prereq FROM R
+  left JOIN course_prerequisites ON course = course_prerequisites.course_id
+  left JOIN prerequisite_course_sets ON course_prerequisites.set_number = prerequisite_course_sets.set_number
+  ORDER BY course, set`;
+
+  if (planCourses.length === 0) {
+    throw createHttpError.NotFound(`No plan with ID ${req.params.planId}`);
+  }
+  const plan: CourseInfo[] = [];
+  let coursenum: number = planCourses[0].course;
+  let prevset: number = planCourses[0].set;
+  let prereqGroup: { needed: number; courses: number[] }[] = [];
+  let prereqs: number[] = [];
+  let courseInfo: CourseInfo = {
+    course: coursenum,
+    prerequisites: prereqGroup,
+  };
+  planCourses.forEach((element) => {
+    if (element.course !== coursenum) {
+      prereqGroup.push({
+        needed: prereqs.length,
+        courses: prereqs,
+      });
+      courseInfo.prerequisites = prereqGroup;
+      plan.push(courseInfo);
+      coursenum = element.course;
+      prereqGroup = [];
+      prereqs = [];
+      prevset = element.set;
```

Comment:
Instead of keeping track of a `prevset` variable, can we replace the `.forEach()` with a standard for loop, and just refer to the previous index?

Reply:
I'd rather keep working with the forEach, I'd have to add specific tracking for the first index anyways since trying to go negative would cause funkyness anyways, so I don't think it would really do much